### ISSUE TITLE
`.code_ptr()` method unsound

### ISSUE THREAD

**Issue by Monadic-Cat** [tov/libffi-rs#26]
The [`code_ptr`](https://docs.rs/libffi/1.0.1/libffi/high/arity0/struct.Closure0.html#method.code_ptr) method on the `ClosureN` structs in the `high` interface is unsound, as `fn(...) -> R` pointers are `Copy`, so you can copy the returned function pointer out from behind the reference with the appropriate lifetime and "safely" call them after the backing closure has been dropped.

**Comment by Monadic-Cat** [comment]
These two options have been suggested in conversation in the Rust Community server on Discord:
- Make the returned `fn` pointer `unsafe` to call. (e.g., `unsafe extern "C" fn() -> R`)
- Return it inside a `#[repr(transparent)]` wrapper struct to enforce the lifetime
  ```rust
  #[repr(transparent)]
  struct FnPtr0<'a, R> {
     ptr: extern "C" fn() -> R,
     _lifetime: ::core::marker::PhantomData<&'a extern "C" fn() -> R>,
  }
  ```
  It is worth noting that this option would mean that calling the returned function on the Rust side would require going through a distinct `.call()` method on `FnPtrN` to pass through to the `fn` pointer, as implementing the `Fn*` traits is unstable. :(

**Comment by YorickPeterse** [comment]
The `high` interface as a whole is a bit dodgy. I don't use it myself so I haven't really done anything about it. Given this is a high-level interface, I think a wrapper type is best. Making it `unsafe` feels a bit too fragile.

**Comment by Monadic-Cat** [comment]
I might be interested in working on improving the `high` interface, though I don't currently use it either. (I only filed this issue because someone in a Q/A channel I frequent was stumbling over this particular soundness hole.)

**Comment by YorickPeterse** [comment]
Fixed by https://github.com/tov/libffi-rs/pull/27

**Comment by YorickPeterse** [comment]
The fix for this is released in version 2.0.0. Thanks for fixing the soundness hole! :tada:
